pkg/controller/core/workload_controller.go

```diff
@@ -290,6 +291,13 @@ func syncAdmissionCheckConditions(conds []kueue.AdmissionCheckState, queueChecks
 	return conds, shouldUpdate
 }
 
+func (r *WorkloadReconciler) removeFinalizer(ctx context.Context, wl *kueue.Workload) error {
+	if controllerutil.RemoveFinalizer(wl, kueue.ResourceInUseFinalizerName) {
+		return r.client.Update(ctx, wl)
+	}
+	return nil
+}
```

Comment:
There's a similar method in `pkg/controller/jobframework/reconciler.go`. 
https://github.com/kubernetes-sigs/kueue/blob/e12a8717c453fcce962522713bb4239abd2d0059/pkg/controller/jobframework/reconciler.go#L661-L666

We could convert that method into an util function and use it in both places.

Reply:
I would hold from this, as I believe that the fix should be in the reconciler only https://github.com/kubernetes-sigs/kueue/issues/1450#issuecomment-1856148578

Reply:
Alright, I just pushed a change which deletes the finalizer in the case that the stopJob results in the error that the object is not found. In any other case (and failure of removing the finalizer) results in returning an error on reconcile.

Reply:
Uhmmm is the not found the case that you faced in production?
If somehow removing the finalizer fails, I think the next reconcile wouldn't be able to reach to this piece of the code, would it?
Maybe a better solution is closer what you had initially. But we shouldn't try to remove finalizers from two places. So perhaps just doing it in the workload controller only is better.
Whatever you do, add a unit test.

Reply:
Sorry for the back and forth, but I think I have a better understanding of the problem now

Reply:
Yes, the problem I faced is that the Pod is already deleted, but the finalizer is still on the workload resource. 

Alright, let me think of an alternative implementation. If meanwhile you have some ideas, please let me know.

Reply:
To summarize, I think the solution should be:
- Do finalizer removal from Workload controller
- Do NOT do any workload finalizer removal from the job reconciler, for better separation of concerns and to avoid the controllers from stumbling into each other.